src/proposals.rs

```diff
@@ -177,10 +186,32 @@ async fn poll_approved_proposals(
                 proposal_id
             );
             state.increment_proposal_id();
-            continue;
         }
 
         let proposal = proposal.unwrap();
+        match proposal.status() {
+            ProposalStatus::DepositPeriod | ProposalStatus::VotingPeriod => {
+                info!("proposal {} is in deposit or voting period", proposal_id);
+
+                // return without incrementing so this ID will be checked again after a period
+                break;
```

Comment:
Is `break` what we meant here? Isn't this supposed to be an infinite loop? Another thing to keep in mind is that a proposal in `DepositPeriod` may not ever enter `VotingPeriod`, since a full deposit isn't necessary to submit a proposal. If we have to halt when we see one, we'd have to wait the entire deposit period potentially before it could execute the next proposal, after which the block height in a successful one may have already passed.

Reply:
Hrmm, on further reflection I think given that deposit periods are the same length as voting periods, it wouldn't be the case where it never exits the deposit period -- rather someone waiting the max amount of time in deposit, then voting, which would have the effect of 96 hours of block time, during which another proposal would be stuck queued behind it for up to 48 extra hours.

Reply:
The usage of a `break` was meant to allow the polling thread to sleep for `query_period` before checking the same proposal again, whereas `continue` would check it immediately. It is a little janky since the query period is configurable and could be short.

Can't anyone deposit to a proposal? In the short term we could prevent it through that means; maybe a more robust thread scheduling system for handling proposals in different states is needed? Do we need to account for proposals that may depend on state from previous cork proposals?

Reply:
I don't think we'd want to get into the business of paying deposits on bad gov props -- it's also a very cheap attack to create a lot of these. The most straightforward solution here in my mind is this:

1. Set the last observed prop ID to the highest value where there are no intermediate states before it (everything is passed/rejected/failed/missing) -- so we'd only increment on observing those states and not increment past if we see anything in deposit/voting period.
2. Reprocess everything after that ID, but keep a hash map of prop IDs we've already submitted to so we don't have to re-do it. They should be idempotent, but without this it will end up with a bunch of TX spam on the chain.
3. We don't need to care about dependency on prior state -- in the end it's up to submitters to time out and schedule when they want these things to execute. Bear in mind that a passed prop cork has an execution height, it will execute then rather than when the proposal is approved.